web3.js/src/connection.ts

```diff
@@ -2761,80 +2762,257 @@ export class Connection {
+  //   try {
+  //     decodedSignature = bs58.decode(signature);
+  //   } catch (err) {
+  //     throw new Error('signature must be base58 encoded: ' + signature);
+  //   }
+
+  //   assert(decodedSignature.length === 64, 'signature has invalid length');
+
+  //   const subscriptionCommitment = commitment || this.commitment;
+  //   let subscriptionId;
+  //   let done = false;
+  //   let response: RpcResponseAndContext<SignatureResult> | null = null;
+
+  //   const confirmTx = new Promise((resolve, reject) => {
+  //     try {
+  //       subscriptionId = this.onSignature(
+  //         signature,
+  //         (result: SignatureResult, context: Context) => {
+  //           subscriptionId = undefined;
+  //           response = {
+  //             context,
+  //             value: result,
+  //           };
+  //           done = true;
+  //           resolve(TransactionStatus.CONFIRMED);
+  //         },
+  //         subscriptionCommitment,
+  //       );
+  //     } catch (err) {
+  //       reject(err);
+  //     }
+  //   });
+
+  //   const checkBlockHeight = async () => {
+  //     const blockHeight = await this.getBlockHeight(commitment);
+  //     return blockHeight;
+  //   };
+
+  //   const expireTx = new Promise((resolve, reject) => {
+  //     (async () => {
+  //       try {
+  //         let currentBlockHeight = await checkBlockHeight();
+  //         while (currentBlockHeight <= latestBlockhash.lastValidBlockHeight!) {
+  //           if (done) return;
+  //           await sleep(1000);
+  //           currentBlockHeight = await checkBlockHeight();
+  //         }
+  //         resolve(TransactionStatus.EXPIRED);
+  //       } catch (error) {
+  //         reject(error);
+  //       }
+  //     })();
+  //   });
+
+  //   try {
+  //     const outcome = await Promise.race([confirmTx, expireTx]);
+  //     switch (outcome) {
+  //       case 'CONFIRMED':
+  //         console.log('Transaction has been confirmed');
+  //         break;
+  //       case 'EXPIRED':
+  //         console.log('Transaction has expired');
+  //         break;
+  //     }
+  //   } finally {
+  //     if (subscriptionId) {
+  //       this.removeSignatureListener(subscriptionId);
+  //     }
+  //   }
+
+  //   if (response === null) {
+  //     throw new Error(`Transaction ${signature} expired`);
+  //   }
+
+  //   return response;
+  // }
+
+  confirmTransaction(
+    signature: {
+      signature: TransactionSignature;
+      latestBlockhash: TransactionBlockhash;
+    },
+    commitment?: Commitment,
+  ): Promise<RpcResponseAndContext<SignatureResult>>;
+
   /**
-   * Confirm the transaction identified by the specified signature.
+   * @deprecated Using confirmTransaction with only a signature is deprecated.
    */
-  async confirmTransaction(
+  confirmTransaction(
     signature: TransactionSignature,
     commitment?: Commitment,
+  ): Promise<RpcResponseAndContext<SignatureResult>>;
+
+  async confirmTransaction(
+    signature: any,
```

Comment:
I think this is reasonable. I had `any` because the examples for writing overloads I had found all said that the types on the implementation needed to be as generic as possible to be able to match the overloads. But in the `Transaction` constructor overload we had it as a union of possible types and that worked. I'll get this changed.